server.js

```diff
@@ -1,38 +1,167 @@
+    await hufflepuff.save()
+
+    // await new Book({ title: "Harry Potter and the Philosopher's Stone", house: gryffindor }).save()
+    // await new Book({ title: "Harry Potter and the Chamber of Secrets", house: slyterhin }).save()
+
+    potterData.forEach((character) => {
+      new Character(character).save()
+    })
+  }
+  seedDatabase()
+}
+
 // Defines the port the app will run on. Defaults to 8080, but can be 
 // overridden when starting the server. For example:
-//
 //   PORT=9000 npm start
-const port = process.env.PORT || 8080
+const port = process.env.PORT || 9090
 const app = express()
 
-// Add middlewares to enable cors and json body parsing
+// MIDDLEWARES
 app.use(cors())
 app.use(bodyParser.json())
+//A function that recives a request, the response and the argument "next". 
+//It will execute before the routes below, if it's not evoked next() it will block the code coming next. The connection.readyState checks that the connection is good.
+app.use((req, res, next) => {
+  if (mongoose.connection.readyState === 1) {
+    next()
+  } else {
+    res.status(503).json({ error: 'Service unavailable' })
+  }
+})
 
-// Start defining your routes here
+// GET ROUTES
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send('Mongo-project: Harry Potter characters')
+})
+
+//GET ALL HOUSES. How do I combine the house model with the character model? 
+app.get('/houses', async (req, res) => {
+  const houses = await House.find()
+  res.json(houses)
+})
+
+//Trying to get the house information populated into the character model but  can't make it to work. 
+//Book is to test if it works with only "manually" created data and that worked... http://localhost:9090/books
+// app.get('/books', async (req, res) => {
+//   const books = await Book.find().populate('house')
+//   res.json(books)
+// })
+
+//GET ALL CHARACTERS, Filter on name, house and job. http://localhost:9090/characters?name=harry&job=student&house=gryffindor&gender=male
+app.get('/characters', (req, res) => {
+  let queryObj = {}
+  const pageOptions = {
+    page: parseInt(req.query.page, 10) || 0,
+    limit: parseInt(req.query.limit, 10) || 10
+  };
+  //Regular expression to make it case insensitive
+  if (req.query.name) { queryObj['name'] = new RegExp(req.query.name, 'i') }
+  if (req.query.house) { queryObj['house'] = new RegExp(req.query.house, 'i') }
+  if (req.query.job) { queryObj['job'] = new RegExp(req.query.job, 'i') }
+  if (req.query.gender) { queryObj['gender'] = new RegExp(req.query.gender, 'i') }
+
+  Character.find(queryObj)
+    .skip(pageOptions.page * pageOptions.limit)
+    .limit(pageOptions.limit)
+    .sort({ id: 1 })
+    .then((results) => {
+      // Succesfull
+      console.log('Found : ' + results);
+      res.json(results);
+    }).catch((err) => {
+      // Error/Failure
+      console.log('Error ' + err);
+      res.json({ message: 'Cannot find this character', err: err }); //Don't understand when this error message will show?
```

Comment:
This could happen if Mongoose can't complete the search for some reason.  

To handle not finding results, in your `.then()`

One option is to check if there are any results and return something different to the client.
Another option would be to check if there are no results and `throw` and error which would then execute whatever is in `.catch((err) => {...})`